astronet/direct_tensor/predict.py

```diff
@@ -0,0 +1,382 @@
+"""Make astronet predictions without creating tf.training.Example objects and files."""
+
+import json
+from itertools import starmap
+from multiprocessing import Pool
+from pathlib import Path
+from typing import Any, Literal, Optional, Protocol
+
+import numpy as np
+import pandas as pd
+import tensorflow as tf
+
+from astronet.direct_tensor.features import (
+    aperture_features,
+    double_period_features,
+    even_features,
+    global_features,
+    half_period_features,
+    local_features,
+    odd_features,
+    sample_segments_features,
+    secondary_features,
+)
+from astronet.preprocess import preprocess
+
+
+class LCGetter(Protocol):
+    """Take astro_id and optionally aperture name and return (time, flux)."""
+
+    def __call__(
+        self, astro_id: int, aperture: Optional[Literal["s", "m", "l"]] = None
+    ) -> tuple[np.ndarray, np.ndarray]: ...
+
+
+def standard_view_features(
+    tic: int,
+    time: np.ndarray,
+    flux: np.ndarray,
+    period: float,
+    epoch: float,
+    duration: float,
+    breakspace: Optional[float],
+    aperture_fluxes: dict[str, tuple[np.ndarray, np.ndarray]],
+) -> dict[str, Any]:
+    """
+    Process lightcurve and create standard view inputs that depend on time/flux values.
+
+    Params
+    ------
+    tic: int
+        TIC ID of target. Note: pretty sure this is not used by any of the methods where it's used.
+    time: array[float]
+        Array of time values in lightcurve.
+    flux: array[float]
+        Array of flux values in lightcurve.
+    period: float
+        Period of transit signal; used for lightcurve folding.
+    epoch: float
+        Reference transit time for signal; used for lightcurve folding.
+    duration: float
+        Duration of transits.
+    breakspace: float | None
+        Breakspace to use in spline detrending. None indicates that various breakspaces should be
+        tried and the best selected.
+    aperture_fluxes: dict[str, tuple[np.ndarray, np.ndarray]]
+        For triage model: empty dict.
+        For vetting model: dict of {aperture_name: (lightcurve_time, lightcurve_flux)} for all
+        apertures to be considered ('s', 'm', 'l').
+
+
+    Returns
+    -------
+    features: dict[str, float | array[float]]
+        Dict of {feature_name: feature_value} for features that are or depend on views of the
+        lightcurve.
+    fold_num: array[float]
+        Array containing the fold number of each point in the lightcurve. Used to calculate the
+        total number of folds in the lightcurve for the "n_folds" feature.
+    """
+    tag = "" if breakspace is None else f"_{breakspace}"
```

Comment:
Oh, this one is important, at least in my understanding of the astronet model on PDO. The features are expected in the format of `{feature_name}{tag}` or maybe `{feature_name}.{tag}`. I think it's the former.

Reply:
Since it's a common format, I would write a transformer which takes the features and updates the keys using the `tag` (or just pass the tag into the feature generation functions).

Reply:
The tag is correct, but move it closer to where it's used (at the end of the function). Also, tensorflow doesn't like periods in feature names, so do that normalization here instead of letting tensorflow do it.